### PR TITLE
Add json module

### PR DESCRIPTION
**Pull request by michalmuskala**
This implements the `json` module as specified in [EEP 68](https://github.com/erlang/eep/pull/59).

### CODE REVIEW

File: lib/stdlib/src/json.erl [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+-spec encode(simple_encode_value()) -> iodata().
+encode(Term) -> encode(Term, fun json:encode_value/2).
+
+-doc "
+Generates JSON corresponding to `Term`.
+
+Can by customised with the `Encoder` callback.
+The callback will be recursively called for all the data
+to be encoded and is expected to return the corresponding
+encded values.
```
Comment (elbrujohalcon) [inline lib/stdlib/src/json.erl:114]:
```suggestion
encoded values.
```

File: lib/stdlib/src/json.erl [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+-doc "
+Generates JSON corresponding to `Term`.
+
+Can be customised with the `Encoder` callback.
+The callback will be recursively called for all the data
+to be encoded and is expected to return the corresponding
+encoded values.
+
+Various `encode_*` functions in this module can be used
+to help in construcing such callbacks.
```
Comment (ilya-klyuchnikov) [inline lib/stdlib/src/json.erl:117]: construcing -> constructing

File: lib/stdlib/src/json.erl [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+
+Supports basic data mapping:
+| **JSON** | **Erlang**           |
+|----------|----------------------|
+| Number   | integer() \| float() |
+| Boolean  | true \| false        |
+| Null     | null                 |
+| String   | binary()             |
+| Object   | #{binary() => _}     |
+".
```
Comment (dgud) [inline lib/stdlib/src/json.erl:392]: Does not render a table, [decode doc](https://erlang.github.io/prs/8111/lib/stdlib-5.2/doc/html/json.html#decode/1)

File: lib/stdlib/src/json.erl [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+Any left-over, unparsed data in `Binary` will be returned.
+
+## Example
+
+Decoding object keys as atoms:
+
+```erlang
+Push = fun(Key, Value, Acc) -> [{binary_to_existing_atom(Key), Value} | Acc] end,
+json:decode(Data, #{object_push => Push})
+```
```
Comment (dgud) [inline lib/stdlib/src/json.erl:417]: There is no `json:decode/2`.
This example (and default code) also assumes the *Acc* is a list, should probably be documented.

File: lib/stdlib/test/json_SUITE.erl [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+decode(Bin) -> json:decode(Bin).
+
+%%
+%% JSON SUITE tests
+%%
+
+test_json_test_suite(Config) ->
+    DataDir = proplists:get_value(data_dir, Config),
+    TestFiles = filelib:wildcard("*.json", DataDir),
```
Comment (dgud) [inline lib/stdlib/test/json_SUITE.erl:412]: Testfiles not included in this PR, should it be, what license?

**Review by dgud — COMMENTED** [review]
Some initial comments
Docs are a bit sparse :-)
Misses meta-data since 27.

But overall looks good to me.

File: lib/stdlib/src/json.erl [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+decode(Binary) when is_binary(Binary) ->
+    case value(Binary, Binary, 0, ok, [], #decode{}) of
+        {Result, _Acc, <<>>} -> Result;
+        {_, _, Rest} -> unexpected(Rest, 0)
+    end.
+
+-doc "
+Parses a JSON value from `Binary`.
+
+Similar to `decode/1` escept the decoding process
```
Comment (patrickcarver) [inline lib/stdlib/src/json.erl:403]: escept -> except

File: lib/stdlib/test/json_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+    property_integer_roundtrip/1,
+    property_float_roundtrip/1,
+    property_object_roundtrip/1,
+    property_escape_all/1
+]).
+
+-define(PROPER_NO_TRANS, true).
+-define(PROPER_NO_IMPORT_PARSE, true).
+-include_lib("proper/include/proper.hrl").
+-endif.
```
Comment (juhlig) [inline lib/stdlib/test/json_SUITE.erl:64]: AFAIK, property tests are usually put in their own suite, using the `ct_property_test` framework as described [here](https://www.erlang.org/doc/apps/common_test/ct_property_test_chapter). At least all the property tests for other modules in OTP that I know of are organized this way. For example, see [`stdlib/test/base64_property_test_SUITE.erl`](https://github.com/erlang/otp/blob/master/lib/stdlib/test/base64_property_test_SUITE.erl) and the corresponding [`stdlib/test/property_test/base64_prop.erl`](https://github.com/erlang/otp/blob/master/lib/stdlib/test/property_test/base64_prop.erl).

**Review by juhlig — CHANGES_REQUESTED** [review]

File: lib/stdlib/src/json.erl [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+        true ->
+            error({duplicate_key, Key});
+        _ ->
+            Visited = Visited0#{EncodedKey => true},
+            [$,, EncodedKey, $:, Encode(Value, Encode) | do_encode_checked(Rest, Encode, Visited)]
+    end;
+do_encode_checked([], _, _) ->
+    [].
+
+%% Dispatching through `Encode` any value could allow incorrect
```
Comment (paulo-ferraz-oliveira) [inline lib/stdlib/src/json.erl:199]:
```suggestion
%% Dispatching "any value" through `Encode` could allow incorrect
```

File: lib/stdlib/src/json.erl [hunk 9]
```diff
@@ -0,0 +1,765 @@
...
+
+-doc "
+Parses a JSON value from `Binary`.
+
+Similar to `decode/1` escept the decoding process
+can be customized with the callbacks specifified in
+`Decoders`. The callbacks will use the `Acc` value
+as the initial accumulator.
+
+Any left-over, unparsed data in `Binary` will be returned.
```
Comment (paulo-ferraz-oliveira) [inline lib/stdlib/src/json.erl:408]:
```suggestion
Any leftover unparsed data, in `Binary`, will be returned.
```

File: lib/stdlib/src/json.erl [hunk 10]
```diff
@@ -0,0 +1,765 @@
...
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%% % @format
+%%
+-module(json).
+
```
[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply lib/stdlib/src/json.erl:22]: Also some more examples, maybe (?)

File: lib/stdlib/src/json.erl [hunk 11]
```diff
@@ -0,0 +1,765 @@
...
+    Decode#decode{float = Fun};
+parse_decoder(integer, Fun, Decode) when is_function(Fun, 1) ->
+    Decode#decode{integer = Fun};
+parse_decoder(string, Fun, Decode) when is_function(Fun, 1) ->
+    Decode#decode{string = Fun};
+parse_decoder(null, Null, Decode) ->
+    Decode#decode{null = Null}.
+
+value(<<Byte, Rest/bits>>, Original, Skip, Acc, Stack, Decode) ->
+    Byte = binary:at(Original, Skip),
```
[the remark this answers is not in the document]
Comment (michalmuskala) [reply lib/stdlib/src/json.erl:451]: Yes, sorry. Left over from debugging

**Comment by michalmuskala** [comment]
This changes the string escaping implementation to read multiple bytes at once, and applies several other encoding optimisations. With those changes, in my benchmarks on an M1 mac, the new `json` module significantly outperforms `Jason` by 1.5-2.5x, and `jiffy` (except for one case where it's on-par).
This includes implementing custom UTF8 validation that significantly outperforms `/utf8` patterns in case where we need to know the length of the codepoint, but don't need to know its value.

Btw, in the implementation of UTF8 validation using tuples and `element` calls to do lookups, rather than functions with integer patterns seems to be considerably faster - I believe this is because of optimisations the compiler tries to do by fusing certain clauses, which in this case leads to a pesimisation.

Here's direct reference to the commit with latest changes: https://gist.github.com/michalmuskala/3e709dc20839bf5159e7fd967e46d0bf

File: lib/stdlib/src/json.hrl [hunk 12]
```diff
@@ -0,0 +1,144 @@
+
+-define(is_ascii_escape(Byte),
```
Comment (dkuku) [inline lib/stdlib/src/json.hrl:2]: Wouldn't checking the range and just the two outliers be more performant and easier to read ?

File: lib/stdlib/src/json.erl [hunk 13]
```diff
@@ -0,0 +1,864 @@
...
+-doc "
+Generates JSON corresponding to `Term`.
+
+Can be customised with the `Encoder` callback.
+The callback will be recursively called for all the data
+to be encoded and is expected to return the corresponding
+encoded values.
+
+Various `encode_*` functions in this module can be used
+to help in constructing such callbacks.
```
Comment (codeadict) [inline lib/stdlib/src/json.erl:122]:
```suggestion
to help constructing such callbacks.
```

**Comment by josevalim** [comment]
> This includes implementing custom UTF8 validation that significantly outperforms /utf8 patterns in case where we need to know the length of the codepoint, but don't need to know its value.

I can see such a function being useful in other scenarios. Perhaps it makes sense to export it as part of `unicode`?

File: lib/stdlib/src/json.erl [hunk 14]
```diff
@@ -0,0 +1,864 @@
...
+-define(UTF8_ACCEPT, 0).
+-define(UTF8_REJECT, 12).
+
+%%
+%% Encoding implementation
+%%
+
+-type encoder() :: fun((dynamic(), encoder()) -> iodata()).
+
+-type simple_encode_value() ::
```
Comment (seriyps) [inline lib/stdlib/src/json.erl:81]: I see how in the future many will want to use this type in their own typespecs as "the JSON type". Might it make sense to try to make the name shorter?

File: lib/stdlib/src/json.erl (hunk 13 above)
Comment (sneako) [reply lib/stdlib/src/json.erl:122]: Respectfully disagree, I believe the "in" is correct here

File: lib/stdlib/src/json.erl [hunk 15]
```diff
@@ -0,0 +1,864 @@
...
+Can be customised with the `Encoder` callback.
+The callback will be recursively called for all the data
+to be encoded and is expected to return the corresponding
+encoded values.
+
+Various `encode_*` functions in this module can be used
+to help in constructing such callbacks.
+".
+-spec encode(dynamic(), encoder()) -> iodata().
+encode(Term, Encoder) when is_function(Encoder, 2) ->
```
[the remark this answers is not in the document]
Comment (michalmuskala) [reply lib/stdlib/src/json.erl:125]: This is proposed in the EEP, but not implemented yet - to be done as a follow-up.

See https://github.com/michalmuskala/eep/blob/json-eep/eeps/eep-0068.md#formatting-and-pretty-printing for details from the EEP

**Comment by michalmuskala** [comment]
> > This includes implementing custom UTF8 validation that significantly outperforms /utf8 patterns in case where we need to know the length of the codepoint, but don't need to know its value.
> 
> I can see such a function being useful in other scenarios. Perhaps it makes sense to export it as part of `unicode`?

Unfortunately this function being efficient in here very much depends on it being placed in the same module as the caller allowing the compiler to analyse and optimise them together. I'm not convinced this would be particularly efficient as a standalone function. I'd prefer to focus on further optimising `/utf8` patterns, so the "intuitive" way of doing things is also the fastest.

File: lib/stdlib/src/json.erl (hunk 13 above)
Comment (danilagamma) [reply lib/stdlib/src/json.erl:122]: I believe "with" should be used instead of "in".

**Comment by codeadict** [comment]
> > This includes implementing custom UTF8 validation that significantly outperforms /utf8 patterns in case where we need to know the length of the codepoint, but don't need to know its value.
> 
> I can see such a function being useful in other scenarios. Perhaps it makes sense to export it as part of `unicode`?

Would this help with UTF-8 validation in Cowboy and Bandit?

File: lib/stdlib/src/json.hrl (hunk 12 above)
Comment (michalmuskala) [reply lib/stdlib/src/json.hrl:18]: No, this implementation is significantly more performant. The compiler & JIT generate an efficient jump table in this case that is also fused with subsequent calls, e.g. to `escape/1`. Range comparisons do a lot of repeated work.

File: lib/stdlib/src/json.erl (hunk 14 above)
Comment (michalmuskala) [reply lib/stdlib/src/json.erl:82]: That's a good question - I'm going to rename those types to just `encode_value` and `decode_value`

File: lib/stdlib/src/json.erl (hunk 10 above)
[the remark this answers is not in the document]
Comment (michalmuskala) [reply lib/stdlib/src/json.erl:32]: I pushed a commit with more documentation & more examples

File: lib/stdlib/test/json_SUITE.erl (hunk 5 above)
Comment (michalmuskala) [reply lib/stdlib/test/json_SUITE.erl:564]: This was added now. The files are MIT-licensed - I added such a note in a README file with the folder with test files, as well as a link to the original source.

**Comment by michalmuskala** [comment]
I've pushed commits that tune the performance of decoder (and tiny changes to encoder). This also includes the change to decoder to remove the `empty_object` and `empty_array` settings in favour of calling callbacks, e.g. `ArrayFinish(ArrayStart(Acc))`.

Current benchmark results for an M1 mac can be found in https://gist.github.com/michalmuskala/00e615f5de4305027abecffed45af07b

The decoder is the fastest out of all libraries I've checked, except for `jiffy` in couple specific data sets (notably lots of tiny strings and unicode-heavy data). In other cases the decoder is usually ~30% faster than `Jason` and `jiffy` (and up to 2x faster) and 2x faster than `jsone` (and up to 5x faster).

I've also improved documentation and added some more examples.

**Comment by michalmuskala** [comment]
> Would this help with UTF-8 validation in Cowboy and Bandit?

cowboy already does this - this is actually how I learned about this technique, though I'm pretty sure my implementation is more performant.

https://github.com/ninenines/cowlib/blob/cc04201c1d0e1d5603cd1cde037ab729b192634c/src/cow_ws.erl#L581-L588

File: lib/stdlib/src/json.erl [hunk 16]
```diff
@@ -0,0 +1,864 @@
...
+-spec encode_map(map(), encoder()) -> iodata().
+encode_map(Map, Encode) when is_map(Map) ->
+    do_encode_map(Map, Encode).
+
+do_encode_map(Map, Encode) when is_function(Encode, 2) ->
+    encode_object([[$,, key(Key, Encode), $: | Encode(Value, Encode)] || Key := Value <- Map]).
+
+-spec encode_map_checked(map(), encoder()) -> iodata().
+encode_map_checked(Map, Encode) ->
+    do_encode_checked(maps:to_list(Map), Encode).
```
[the remark this answers is not in the document]
Comment (michalmuskala) [reply lib/stdlib/src/json.erl:254]: It would be possible, but it would mean we can't use the same loop function for kv-lists and maps.

File: lib/stdlib/src/json.hrl [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+%%
+
+%% A lot of the macros below use multi-value comparisons where
+%% range checks would have worked just fine. This is because
+%% the compiler & JIT can emit better code in some cases when
+%% multiple clauses are to be dispatched based on such sets
+%% of values. They'll generate an efficient "jump table",
+%% which gets to the correct clause in one go, rather
+%% than going through a set of comparisons.
+%% However, this migth not always be the bext way (see is_0_to_9),
```
Comment (jparise) [inline lib/stdlib/src/json.hrl:28]:
```suggestion
%% However, this might not always be the best way (see is_0_to_9),
```

File: lib/stdlib/src/json.hrl [hunk 18]
```diff
@@ -0,0 +1,189 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2001-2016. All Rights Reserved.
```
Comment (jparise) [inline lib/stdlib/src/json.hrl:4]: 2024?

File: lib/stdlib/src/json.erl [hunk 19]
```diff
@@ -0,0 +1,1085 @@
...
+-type encode_map(Value) :: #{binary() | atom() | integer() => Value}.
+
+-doc """
+Generates JSON corresponding to `Term`.
+
+Supports basic data mapping:
+
+| **Erlang**              | **JSON** |
+|-------------------------|----------|
+| `integer() \\| float()` | Number   |
```
Comment (garazdawi) [inline lib/stdlib/src/json.erl:115]:
```suggestion
| `integer() \| float()` | Number   |
```

Since you are using `"""` there is no need to escape `\`.

File: lib/stdlib/src/json.erl [hunk 20]
```diff
@@ -0,0 +1,1085 @@
...
+        0 ->
+            escape_binary_ascii(Rest, [Acc | Escape], Orig, Skip, 0);
+        _ ->
+            Part = binary_part(Orig, Skip0, Len),
+            escape_binary_ascii(Rest, [Acc, Part | Escape], Orig, Skip, 0)
+    end;
+escape_binary(<<Byte, Rest/binary>>, Acc, Orig, Skip, Len) ->
+    case element(Byte - 127, utf8s0()) of
+        ?UTF8_REJECT -> invalid_byte(Orig, Skip + Len);
+        %% all accept cases are ASCII, already covred above
```
Comment (Gazler) [inline lib/stdlib/src/json.erl:363]: s/covred/covered

**Comment by michalmuskala** [comment]
I've squashed the commits and rebased on top of latest master

**Comment by lpil** [comment]
Congratulations and thank you @michalmuskala!

**Comment by gonzalobf** [comment]
> This changes the string escaping implementation to read multiple bytes at once, and applies several other encoding optimisations. With those changes, [in my benchmarks on an M1 mac](https://gist.github.com/michalmuskala/3e709dc20839bf5159e7fd967e46d0bf), the new `json` module significantly outperforms `Jason` by 1.5-2.5x, and `jiffy` (except for one case where it's on-par). This includes implementing custom UTF8 validation that significantly outperforms `/utf8` patterns in case where we need to know the length of the codepoint, but don't need to know its value. Unfortunately constructing binaries with `private_append` turned out to still be slower than building an IOlist of parts - even if benchmarked with a final `iolist_to_binary` call.

Hey @michalmuskala,

Great work! Could you please share with us the code of the benchmark you ran?

Thanks

**Comment by michalmuskala** [comment]
This was effectively using benchmarks from the `Jason` project with extra jobs added to check the built-in `json` module (see [here](https://github.com/michalmuskala/jason?tab=readme-ov-file#benchmarks)).

**Comment by essen** [comment]
> Btw, in the implementation of UTF8 validation using tuples and `element` calls to do lookups, rather than functions with integer patterns seems to be considerably faster - I believe this is because of optimisations the compiler tries to do by fusing certain clauses, which in this case leads to a pesimisation.

I'm working on optimising UTF-8 validation in Cowboy Websocket now. I don't believe this to be true, or at least not entirely.

The biggest improvements so far have been from separating plain-text from UTF-8, and making plain-text validating 4 bytes at a time when possible.

Then I also simplified the UTF-8 validation function to reduce the number of tests (instead of `C >= NUM1, C < NUM2` I just do `C < NUM2`). But that doesn't seem to have a big impact.

Using tables shows no significant difference either. However I do get a slight improvement when I modify `utf8t()` by dropping the first 128 bytes - since plain-text characters were handled earlier. The lookup ending up being `element(C - 127, utf8t())` instead of `C + 1`. I think you're doing this with `utf8s0()` but you may not have realised you can do this in all cases - as long as you handle plain-text separately (when you get state 0 you switch to a plain-text specific function).

Of course the implementation is simpler and closer to the source material so unless I find some other way to speed things up the tables may be the way to go.

**Comment by essen** [comment]
OK I see a larger difference with data that's mostly non-plain text that I hadn't noticed before. Nothing huge but significant enough that the tables are definitely the better solution. Thanks for the tip!!

tl;dr See if dropping the first 128 bytes of `utf8t()` helps in your benchmarks.

**Comment by essen** [comment]
One final note (for today). I seem to be getting better results by keeping the table for character classes and implementing separate functions for the different states. Code grows as a result but I get 2 to 4% better performance.

**Comment by essen** [comment]
@michalmuskala Resulting work is in https://github.com/ninenines/cowlib/pull/141/commits/921b1ff6178bd6710331609f226789e9a763e1ae

Am curious if some of it could be applied to `json` but I have other parts of Websocket to optimise now.

**Comment by michalmuskala** [comment]
Thanks for keeping track of this! I'll take a look at what we can bring back
